src/Features/CSharp/Portable/UseRecursivePatterns/CSharpAnalyzedNodeRewriter.cs

```diff
@@ -0,0 +1,397 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseRecursivePatterns
+{
+    using static AnalyzedNode;
+    using static SyntaxFactory;
+
+    internal static class CSharpAnalyzedNodeRewriter
+    {
+        public static ExpressionSyntax AsExpressionSyntax(this AnalyzedNode node)
+        {
+            Debug.Assert(!(node is Evaluation e) || e.Syntax is ExpressionSyntax, "!(node is Evaluation e) || e.Syntax is ExpressionSyntax");
+
+            return node switch
+            {
+                Evaluation p => (ExpressionSyntax)p.Syntax!,
```

Comment:
To avoid doing this we either should run a separate pass to check types. Or return null and run the rewriter before registering the refactoring. I suspect that would be an expensive thing to do for each run of the analysis?